src/main/java/com/unascribed/fabrication/mixin/i_woina/end_portal_parallax/MixinEndPortalRenderer.java

```diff
@@ -0,0 +1,178 @@
+import com.unascribed.fabrication.support.EligibleIf;
+import com.unascribed.fabrication.support.Env;
+import com.unascribed.fabrication.support.SpecialEligibility;
+import com.unascribed.fabrication.support.injection.FabInject;
+import net.minecraft.block.entity.EndPortalBlockEntity;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.render.VertexConsumer;
+import net.minecraft.client.render.block.entity.EndPortalBlockEntityRenderer;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.Util;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.util.math.Matrix4f;
+import org.lwjgl.BufferUtils;
+import org.lwjgl.opengl.GL11;
+import org.lwjgl.system.MemoryStack;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import java.nio.FloatBuffer;
+import java.nio.IntBuffer;
+import java.util.Random;
+
+@Mixin(EndPortalBlockEntityRenderer.class)
+@EligibleIf(configAvailable="*.end_portal_parallax", envMatches=Env.CLIENT, specialConditions=SpecialEligibility.NOT_MACOS)
+public abstract class MixinEndPortalRenderer {
+
+	@Shadow
+	@Final
+	public static Identifier PORTAL_TEXTURE;
+	@Shadow
+	@Final
+	public static Identifier SKY_TEXTURE;
+	@Unique
+	private static final Random fabrication$random = new Random(31100L);
+	private static final FloatBuffer fabrication$matrix_buffer = BufferUtils.createFloatBuffer(4*4);
+
+	@FabInject(at=@At("HEAD"), method= "renderSide(Lnet/minecraft/block/entity/EndPortalBlockEntity;Lnet/minecraft/util/math/Matrix4f;Lnet/minecraft/client/render/VertexConsumer;FFFFFFFFLnet/minecraft/util/math/Direction;)V", cancellable=true)
+	public void oldRender(EndPortalBlockEntity be, Matrix4f model, VertexConsumer vertices, float x1, float x2, float y1, float y2, float z1, float z2, float z3, float z4, Direction side, CallbackInfo ci) {
+		if (!FabConf.isEnabled("*.end_portal_parallax")) return;
+		ci.cancel();
+		if (!be.shouldDrawSide(side)) return;
+		MinecraftClient mc = MinecraftClient.getInstance();
+		try (MemoryStack stack = MemoryStack.stackPush()) {
+			IntBuffer viewportBuffer = stack.mallocInt(4);
+			FloatBuffer scratchBuffer = stack.mallocFloat(16);
+			FloatBuffer objectCoords = stack.mallocFloat(3);
+			FloatBuffer projBuffer = stack.mallocFloat(16);
+
+			GL11.glGetIntegerv(GL11.GL_VIEWPORT, viewportBuffer);
+			Matrix4f scratchMat = RenderSystem.getModelViewMatrix().copy();
+			Matrix4f projMatrix = RenderSystem.getProjectionMatrix().copy();
+			scratchMat.multiply(model);
+			projMatrix.writeRowMajor(projBuffer);
+			scratchMat.writeRowMajor(scratchBuffer);
+
+			float objectX = objectCoords.get(0);
+			float objectY = objectCoords.get(1);
+			float objectZ = objectCoords.get(2);
```

Comment:
objectCoords is not assigned anywhere. This won't render correctly, and due to the usage of `malloc` rather than `calloc`, will use random garbage memory as "object coordinates". You need the gluUnProject port.